domainobjs/ts/ballot.ts

```diff
@@ -0,0 +1,127 @@
+import * as assert from "assert"
+import { genRandomSalt, hash5, hashLeftRight, IncrementalQuinTree } from "maci-crypto";
+
+/** 
+ * A Ballot represents a User's votes in a Poll, as well as their next valid
+ * nonce.
+ * @param _voiceCreditBalance The user's voice credit balance
+ * @param _nonce The number of valid commands which the user has already
+ *               published
+*/
+export class Ballot {
+    public votes: bigint[] = []
+    public nonce: bigint = BigInt(0)
+    public voteOptionTreeDepth: number
+
+    /**
+     * Create a new Ballot instance
+     * @param _numVoteOptions How many vote options are available in the poll
+     * @param _voteOptionTreeDepth The depth of the merkle tree holding the vote options
+     */
+    constructor(_numVoteOptions: number, _voteOptionTreeDepth: number) {
+        this.voteOptionTreeDepth = _voteOptionTreeDepth
+        assert(5 ** _voteOptionTreeDepth >= _numVoteOptions)
+        assert(_numVoteOptions >= 0)
+        for (let i = 0; i < _numVoteOptions; i++) {
+            this.votes.push(BigInt(0))
+        }
+    }
+
+    /**
+     * Generate an hash of this ballot
+     * @returns The hash of the ballot
+     */
+    public hash = (): bigint => {
+        const vals = this.asArray()
+        return hashLeftRight(vals[0], vals[1]) as bigint 
+    }
+
+    /**
+     * Convert in a format suitable for the circuit
+     * @returns the ballot as a BigInt array
+     */
+    public asCircuitInputs = (): bigint[] => {
+        return this.asArray()
+    }
+
+    /**
+     * Convert in a an array of bigints
+     * @notice this is the nonce and the root of the vote option tree
+     * @returns the ballot as a bigint array
+     */
+    public asArray = (): bigint[] => {
+        let lastIndexToInsert = this.votes.length - 1
+        while (lastIndexToInsert > 0) {
+            if (this.votes[lastIndexToInsert] !== BigInt(0)) {
+                break
+            }
+            lastIndexToInsert--;
+        }
```

Comment:
Loving your suggestions to make the code more concise, thank you! unfortunately as I rebased loads of these were outdated and GH wouldn't let me commit them from the web UI, so had to add them manually

Reply:
This one however will not work if the votes array is empty. For instance, when initializing a ballot with 0 vote options,  and running the asArray function, the lastIndex will be set to zero; the loop will run at least once because `i = 0; i <= lastIndexToInsert` -> and try to add an undefined value into the tree -> causing this `TypeError: Cannot convert undefined to a BigInt`
Currently, empty ballots are created here to generate the zero values for the incremental merkle trees: https://github.com/privacy-scaling-explorations/maci/blob/dev/contracts/ts/genEmptyBallotRootsContract.ts

What I did, in case the index is not found, we keep -1 so the loop does not run, and in any other cases it runs up until (and including) the last index to include

Reply:
Good catch. So let's use this line instead:
```
const lastIndexToInsert = foundIndex < 0 ? foundIndex : lastIndex - foundIndex
```

Reply:
> This one however will not work if the votes array is empty

@ctrlc03 should this ever be the case? Why would we allow empty user ballots 🤔